### PR TITLE
Refactor Openharmony platform code to support switching between napi (ark js runtime) and v8

### PR DESCRIPTION
**Pull request by dumganhar**
Re: # https://github.com/cocos/3d-tasks/issues/17456

https://github.com/cocos/creator-runtime-extensions/pull/470

https://github.com/cocos/cocos-engine-external/pull/408

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/platform/openharmony/modules/Screen.cpp [hunk 1]
```diff
@@ -69,9 +77,10 @@ Vec4 Screen::getSafeAreaEdge() const {
 bool Screen::isDisplayStats() {
     se::AutoHandleScope hs;
     se::Value ret;
-    char commandBuf[100] = "cc.profiler.isShowingStats();";
+    const char commandBuf[100] = "cc.profiler.isShowingStats();";
     se::ScriptEngine::getInstance()->evalString(commandBuf, 100, &ret);
     return ret.toBoolean();
+    return true;
```
Comment (qiuguohua) [inline native/cocos/platform/openharmony/modules/Screen.cpp]: This return value can be deleted?

Comment (dumganhar) [reply native/cocos/platform/openharmony/modules/Screen.cpp]: Yep, it's a mistake. I will revert it.

File: native/cocos/engine/Engine.cpp [hunk 2]
```diff
@@ -72,29 +72,40 @@
...
+    int height = static_cast<int>(viewSize.height / dpr);
+    windowVal.toObject()->setProperty("innerWidth", se::Value(width));
+    windowVal.toObject()->setProperty("innerHeight", se::Value(height));
+
+    if (sizeof(handler) == 8) { // use bigint
+        windowVal.toObject()->setProperty("windowHandle", se::Value(static_cast<uint64_t>(handler)));
+    } else {
+        windowVal.toObject()->setProperty("windowHandle", se::Value(static_cast<uint32_t>(handler)));
     }
-    se->evalString(ss.str().c_str());
```
Comment (dumganhar) [inline native/cocos/engine/Engine.cpp:97]: Don't use evalString here. OH with NAPI backend doesn't allow this.

File: templates/cmake/openharmony.cmake [hunk 3]
```diff
@@ -11,6 +11,13 @@ macro(cc_openharmony_after_target target_name)
     target_link_libraries(${target_name}
         ${ENGINE_NAME}
     )
+
+    target_link_options(${target_name} PRIVATE -Wl,--gc-sections)
+
+    if("${CMAKE_BUILD_TYPE}" STREQUAL "Release")
+        target_link_options(${target_name} PRIVATE -Wl,-s)
```
Comment (dumganhar) [inline templates/cmake/openharmony.cmake:18]: Strip so in release build. `.so` file size now:

-rw-r--r--@ 1 james  staff  79434848 Jan  1  2009 libcocos-napi-debug.so
-rw-r--r--@ 1 james  staff  13582208 Jan  1  2009 libcocos-napi-release.so
-rw-r--r--@ 1 james  staff  99104128 Jan  1  2009 libcocos-v8-debug.so
-rw-r--r--@ 1 james  staff  26263536 Jan  1  2009 libcocos-v8-release.so

File: native/CMakeLists.txt [hunk 4]
```diff
@@ -113,11 +114,6 @@ if(NOT ANDROID AND NOT IOS AND NOT OHOS AND NOT OPENHARMONY)
     set(USE_WEBVIEW OFF)
 endif()
 
-if(OPENHARMONY)
-    set(USE_SE_V8 OFF)
-    set(USE_SE_NAPI ON)
-endif()
```
Comment (dumganhar) [inline native/CMakeLists.txt:119]: NO NEEDED NOW. editor will pass the cmake flags to us:

Refer to https://github.com/cocos/creator-runtime-extensions/pull/470/files
```
    params.cMakeConfig.USE_SE_V8 = `set(USE_SE_V8 ${openHarmonyOS.useV8 ? 'ON' : 'OFF'})`;
    params.cMakeConfig.USE_SE_NAPI = `set(USE_SE_NAPI ${!openHarmonyOS.useV8 ? 'ON' : 'OFF'})`;
```

File: native/CMakeLists.txt [hunk 5]
```diff
@@ -15,6 +15,7 @@ set(ENGINE_NAME cocos_engine)
 cc_set_if_undefined(CC_DEBUG_FORCE           OFF)
 cc_set_if_undefined(USE_SE_V8                ON)
 cc_set_if_undefined(USE_SE_SM                OFF)
+cc_set_if_undefined(USE_SE_NAPI              OFF)
```
Comment (minggo) [inline native/CMakeLists.txt:18]: So default is use v8 on openharmony?

File: native/cocos/engine/Engine.cpp [hunk 6]
```diff
@@ -72,29 +72,40 @@
...
+        jsbVal.toObject()->setProperty("window", se::Value(windowObj));
+        windowVal.setObject(windowObj, true);
+    }
+
+    int width = static_cast<int>(viewSize.width / dpr);
+    int height = static_cast<int>(viewSize.height / dpr);
+    windowVal.toObject()->setProperty("innerWidth", se::Value(width));
+    windowVal.toObject()->setProperty("innerHeight", se::Value(height));
+
+    if (sizeof(handler) == 8) { // use bigint
```
Comment (minggo) [inline native/cocos/engine/Engine.cpp:104]: Please don't use magic number.

File: native/CMakeLists.txt (hunk 5 above)
Comment (dumganhar) [reply native/CMakeLists.txt:18]: Default is NAPI, Could refer to https://github.com/cocos/creator-runtime-extensions/pull/470.

cc_set_if_undefined means that `if it's not defined, then use this value, if it's defined, use the defined value`.

And for openharmony, both USE_SE_V8 and USE_SE_NAPI are defined. 
The default is USE_SE_V8=OFF, USE_SE_NAPI=ON

File: native/cocos/engine/Engine.cpp (hunk 6 above)
Comment (dumganhar) [reply native/cocos/engine/Engine.cpp:104]: It's original code. 

It's for checking pointer size here. Only 8 and 4 are valid values.

File: native/CMakeLists.txt (hunk 5 above)
Comment (minggo) [reply native/CMakeLists.txt:18]: Ok

**Review by minggo — APPROVED** [review]

File: cocos/core/geometry/geometry-native-ext.ts [hunk 7]
```diff
@@ -87,10 +87,10 @@ const defineAttrFloat = (kls: Constructor, attr: string): void => {
     // __nativeFields__ is defined in jsb_geometry_manual.cpp
     const desc: FieldDesc = (kls as any).__nativeFields__[attr];
     const cacheKey = `_$_${attr}`;
-    if (!window.oh) {
-        // openharmony does not support the console.assert interface at this time.
-        console.assert(desc.fieldSize === 4, `field ${attr} size ${desc.fieldSize}`);
-    }
+
+    // eslint-disable-next-line no-console
+    console.assert(desc.fieldSize === 4, `field ${attr} size ${desc.fieldSize}`);
```
Comment (PPpro) [inline cocos/core/geometry/geometry-native-ext.ts]: can use assert interface exported from 'debug.ts', and remove eslint disable

File: cocos/core/geometry/geometry-native-ext.ts [hunk 8]
```diff
@@ -120,10 +120,10 @@ const defineAttrInt = (kls: Constructor, attr: string): void => {
...
     }
     const cacheKey = `_$_${attr}`;
-    if (!window.oh) {
-        // openharmony does not support the console.assert interface at this time.
-        console.assert(desc.fieldSize === 4, `field ${attr} size ${desc.fieldSize}`);
-    }
+
+    // eslint-disable-next-line no-console
+    console.assert(desc.fieldSize === 4, `field ${attr} size ${desc.fieldSize}`);
+
```
Comment (PPpro) [inline cocos/core/geometry/geometry-native-ext.ts]: the same

File: platforms/native/engine/jsb-loader.js [hunk 9]
```diff
@@ -55,7 +55,7 @@ function downloadScript (url, options, onComplete) {
     if (loadedScripts[url]) return onComplete && onComplete();
 
     download(url, (src, options, onComplete) => {
-        if (window.oh) {
+        if (window.oh && window.scriptEngineType === 'napi') {
```
Comment (PPpro) [inline platforms/native/engine/jsb-loader.js:58]: do we need to add the scriptEngineType defined
https://github.com/cocos/cocos-engine/blob/9f9ee38427b42db7086e689fe568252e06d623ad/%40types/jsb.d.ts#L372-L376

File: platforms/native/engine/jsb-gfx.js [hunk 10]
```diff
@@ -57,8 +57,7 @@ deviceProto.copyTexImagesToTexture = function (texImages, texture, regions) {
 
 const oldDeviceCreateSwapchainFunc = deviceProto.createSwapchain;
 deviceProto.createSwapchain = function (info) {
-    // In openharmony, we need to get the window handle through the jsb interface
-    info.windowHandle = window.oh ? jsb.device.getWindowHandle() : jsbWindow.windowHandler;
+    info.windowHandle = jsbWindow.windowHandle;
```
Comment (PPpro) [inline platforms/native/engine/jsb-gfx.js:60]: can we remove this interface for now ? @qiuguohua 
`jsb.device.getWindowHandle() `

This interface is only for OH platform, should also remove the js binding

File: platforms/native/engine/jsb-loader.js (hunk 9 above)
Comment (dumganhar) [reply platforms/native/engine/jsb-loader.js:58]: It's a property in `global` rather than in `jsb` namespace.

Comment (dumganhar) [reply platforms/native/engine/jsb-loader.js:58]: I think we could not add a new property in global namespace, right?

File: platforms/native/engine/jsb-gfx.js (hunk 10 above)
Comment (dumganhar) [reply platforms/native/engine/jsb-gfx.js:60]: Removed at : https://github.com/cocos/cocos-engine/pull/16093/commits/44f999bcae27ccc36e5b87ec7f26ccc806bf1e38

@qiuguohua Plz review that.

File: cocos/core/geometry/geometry-native-ext.ts (hunk 8 above)
Comment (dumganhar) [reply cocos/core/geometry/geometry-native-ext.ts]: Fixed

File: platforms/native/engine/jsb-loader.js (hunk 9 above)
Comment (PPpro) [reply platforms/native/engine/jsb-loader.js:58]: right

**Review by PPpro — APPROVED** [review]

**Review by qiuguohua — APPROVED** [review]

**Review by PatriceJiang — APPROVED** [review]

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Pull request merged**